src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java

```diff
@@ -518,6 +520,33 @@ private String getEntityId(Object entity) {
 		return null;
 	}
 
+	@Nullable
+	private String getEntityRouting(Object entity) {
```

Comment:
the logic should go into the `ElasticsearchPersistentEntity`. We should add a method `boolean hasJoinFieldProperty()` and `ElasticsearchPersistenProperty getJoinFieldProperty()`. This would be initialized in `org.springframework.data.elasticsearch.core.mapping.SimpleElasticsearchPersistentEntity#addPersistentProperty`. The check if a property matches should check the type to be `JoinField`,  not the annotations.

Here in getEntityRouting we'd get that property (like it's done in  `getEntityVersion()`, then retrieve the `parent`value as Object and use `entityConverter.convertId()` to get the String for the routing value.